tasks/IKF.cpp

```diff
@@ -583,5 +497,130 @@ bool IKF::resetHeading(double heading)
                                             Eigen::AngleAxisd(euler[1], Eigen::Vector3d::UnitY()) *
                                             Eigen::AngleAxisd(euler[2], Eigen::Vector3d::UnitX());
-    return ikf_filter.setAttitude(corrected_attitide);
+    Eigen::Matrix3d cov_attitude = Eigen::Matrix3d::Zero();
+    cov_attitude.topLeftCorner<2,2>() = ikf_filter.getCovariance().topLeftCorner<2,2>();
+    cov_attitude(3,3) = 1.e-06;
+    initializeFilter(corrected_attitide, cov_attitude);
+    return true;
+}
+
+bool IKF::addHeadingOffset(double offset, double variance)
+{
+    base::Vector3d euler = base::getEuler(ikf_filter.getAttitude());
+    Eigen::Quaterniond corrected_attitide = Eigen::AngleAxisd(euler[0] + offset, Eigen::Vector3d::UnitZ()) *
+                                            Eigen::AngleAxisd(euler[1], Eigen::Vector3d::UnitY()) *
+                                            Eigen::AngleAxisd(euler[2], Eigen::Vector3d::UnitX());
+    Eigen::Matrix3d cov_attitude = Eigen::Matrix3d::Zero();
+    cov_attitude.topLeftCorner<2,2>() = ikf_filter.getCovariance().topLeftCorner<2,2>();
+    cov_attitude(2,2) = variance;
+    initializeFilter(corrected_attitide, cov_attitude);
+    return true;
+}
+
+void IKF::initializeFilter(const Eigen::Quaterniond& attitude, const Eigen::Matrix3d& cov_attitude)
+{
+    /******************************************/
+    /** Configuration of the attitude filter **/
+    /******************************************/
+    Eigen::Matrix< double, IKFSTATEVECTORSIZE , 1  > x_0; /** Initial vector state **/
+    Eigen::Matrix3d Ra; /** Measurement noise covariance matrix for accelerometers */
+    Eigen::Matrix3d Rg; /** Measurement noise covariance matrix for gyros */
+    Eigen::Matrix3d Rm; /** Measurement noise covariance matrix for mag */
+    Eigen::Matrix3d Ri; /** Measurement noise covariance matrix for inclinometers */
+    Eigen::Matrix <double, IKFSTATEVECTORSIZE, IKFSTATEVECTORSIZE> P_0; /** Initial covariance matrix **/
+    Eigen::Matrix3d Qbg; /** Noise for the gyros bias instability **/
+    Eigen::Matrix3d Qba; /** Noise for the accelerometers bias instability **/
+    Eigen::Matrix3d Qbi; /** Noise for the inclinometers bias instability **/
+    double sqrtdelta_t = 0.0;
+
+    /*************************/
+    /** Noise configuration **/
+    /*************************/
+    if (config.correction_frequency > accnoise.bandwidth)
+        sqrtdelta_t = sqrt(1.0/accnoise.bandwidth); /** Noise depends on frequency bandwidth **/
+    else
+        sqrtdelta_t = sqrt(1.0/config.correction_frequency); /** Noise depends on frequency bandwidth **/
+
+    Ra = Eigen::Matrix3d::Zero();
+    Ra(0,0) = accnoise.resolution[0] + pow(accnoise.randomwalk[0]/sqrtdelta_t,2);
+    Ra(1,1) = accnoise.resolution[1] + pow(accnoise.randomwalk[1]/sqrtdelta_t,2);
+    Ra(2,2) = accnoise.resolution[2] + pow(accnoise.randomwalk[2]/sqrtdelta_t,2);
```

Comment:
Just a suggestion for readability/maintainability (instead of the previous 4 lines -- and similar for the other initializations):
```
 Ra = (accnoise.resolution + (accnoise.randomwalk/sqrtdelta_t).cwiseAbs2()).asDiagonal();
```